dag_in_context/src/utility/context_of.egg

```diff
@@ -0,0 +1,65 @@
+; We only have context for Exprs, not ListExprs.
+(relation ContextOf (Expr Assumption))
+
+; Base case: InContext means ContextOf
+(rule ((InContext ctx e))
+    ((ContextOf (InContext ctx e) ctx)) :ruleset always-run)
+
+; Error checking - each expr should only have a single context
+; TODO: this may not be true
```

Comment:
I think this is not true considering common subexpr?

Reply:
It may not be true, but not because of common subexpressions. `ContextOf` only contains expressions that have `InContext` nodes within them. For example, consider your example of `(InContext (InFun a) (Num 1))` and `(InContext (InFun b) (Num 1))`. It is not true that `(ContextOf (Num 1) (InFun a))`, since just `(Num 1)` is contextless. However, we do know:
`(ContextOf (InContext (InFun b) (Num 1)) (InFun b))`
`(ContextOf (InContext (InFun a) (Num 1)) (InFun b))`
`(ContextOf (Add (InContext (InFun a) (Num 1)) (InContext (InFun a) (Num 1))))`
and so on.

What should not happen is something like this, where contexts that are not equivalent are mixed:
`(ContextOf (Add (InContext (InFun a) (Num 1)) (InContext (InFun b) (Num 1))))`
That is the motivation for this rule. However, I believe that there are cases where two contexts that have not yet been discovered to be equal are mixed? I am not exactly sure on why two contexts is OK sometimes, Oliver and Ryan were discussing this.